Enhanced support for limits and timeouts.

klizhentas opened the pull request:
- Add support for request size limits
- Add support for bandwidth limits
- Expose keep-alive parameters
- Move to 1.3

Commit:
Enhanced support for limits and timeouts.

* Add support for request size limits
* Add support for bandwidth limits
* Expose keep-alive parameters
* Move to 1.3

bradgignac commented on limit/limiter.go:
```diff
@@ -16,50 +16,104 @@ type Limiter interface {
+// MapperFn takes the request and returns token that corresponds to the  request and the amount of tokens this request is going to consume, e.g.
 // * Client ip rate limiter - token is a client ip, amount is 1 request
-// * Client ip memory limiter - token is a client ip, amount is number of bytes to consume
+// * Client ip bandwitdh limiter - token is a client ip, amount is number of bytes to consume
```
Nitpick: bandwidth not bandwitdh

bradgignac commented on location/httploc/httploc.go:
```diff
@@ -135,16 +177,26 @@ func (l *HttpLocation) GetId() string {
 	}
 }
 
+func (l *HttpLocation) limitRequestSize(req request.Request) (*http.Response, error) {
```
It looks like this method doesn't need to return a response at all.

bradgignac commented on location/httploc/httploc.go:
```diff
@@ -86,17 +116,29 @@ func NewLocationWithOptions(id string, loadBalancer LoadBalancer, o Options) (*H
+func (l *HttpLocation) RoundTrip(req request.Request) (*http.Response, error) {
 	originalRequest := req.GetHttpRequest()
+
+	if re, err := l.limitRequestSize(req); err != nil || err != nil {
```
You're duplicating your error condition on either side of the `||`.

bradgignac replied:
What about changing this call to look like this:

```
if l.isRequestOverLimit(req) {
    return nil, errors.FromStatus(http.StatusRequestEntityTooLarge)
}

...

func (l*HttpLocation) isRequestOverLimit(req request.Request) {
    if l.options.Limits.MaxBodyBytes <= 0 {
        return false
    }
    if (req.GetHttpRequest().ContentLength < l.options.Limit.MaxBodyBytes) {
        return false
    }
    return true
}
```

bradgignac commented:
Left a few comments and suggestions. Still getting familiar with the codebase, so they might not make any sense at all.

klizhentas commented:
@bradgignac Thanks!

Commit:
Changing location level connection options, better control over reads.

* Now the location is responsible of reading the request body.
* The library no longer modifes the body of the original reuqest letting http.Server handle and close it.
* Location supports live updates of the client connection settings and updates the transport.

Commit:
Parsing expressions into failover predicates

* Failover predicates now can be constructed from mini language

Examples:

* RequestMethodEq("GET") - allows to failover only get requests
* IsNetworkError - allows to failover on errors
* RequestMethodEq("GET") && AttemptsLe(2) && (IsNetworkError || ResponseCodeEq(408))
  This predicate allows failover for GET requests with maximum 2 attempts with failover
  triggered on network errors or when upstream returns special http response code 408.

Commit:
Support for updating middleware chains affected by changed options
